Testing tips (#2505)
digital-asset/daml: docs/source/app-dev/app-arch.rst

.. |image0| image:: images/BotFlow.png
   :width: 6.5in
   :height: 3.69444in

Testing
=======

> Testing is fundamental to ensure correctness and favor maintainability.

Author: [False friend](https://translate.google.com/?um=1&ie=UTF-8&hl=en&client=tw-ob#view=home&op=translate&sl=it&tl=en&text=favorire), I believe the term I was looking for was "improve".

> Testing is fundamental to ensure correctness and favor maintainability.

Author: Addressed by https://github.com/digital-asset/daml/pull/2505/commits/3de40913e0eee3e2a83009dc541a455643647ae3

> Testing is usually divided in different categories according to its scope and aim:

Reviewer: ```suggestion
Testing is usually divided into different categories according to its scope and aim:
```

- unit testing verifies single properties of individual components
- integration testing verifies that an aggregation of components behaves as expected
- acceptance testing checks that the overall behavior of a whole system satisfies certain criteria

Both tests in the small scale (unit testing) and large (acceptance testing) tend to be specific to the given component or system under test.

This chapter focuses on providing portable approaches and techniques to perform integration testing between your components and an actual running ledger.

> Test the business logic with a ledger

Reviewer: I don't think you need a new paragraph for each sentence in this section. I'd structure it in three paragraphs:

1. DAML Contracts + DAML Ledger Model encapsulate the bulk of your business logic
2. Mocking the Ledger API pretty much requires you to reimplement a full DAML Ledger so don't.
3. The Sandbox is your friend. 

In 3. you could explain options for getting the ledger into a desired state:
  - Write a scenario
  - Write a special ledger initialisation workflow and automation
  - Use a Postgres sandbox with initial state taken from a backed up database.

Author: I would strongly discourage writing a scenario because of the [limitations of using scenarios with the suggested approach](https://github.com/digital-asset/daml/pull/2505/files/fe13e50529b6718a11b35d913703b2df7fb60959#r313346380).

Author: Where does the Postgres backup idea come from? It's creative but I would not recommend it unless there are some established good practices around it. If there are, please share them with me so that I can add them.

Reviewer: Just from general brainstorming around testing. Defining, and loading the desired ledger state is one of the more cumbersome parts currently so it's on my radar as something to offer solutions for. 

Admittedly, I've never actually tried it.

*************************************

In production, your application is going to interact with a DAML model deployed on an actual ledger.

> Each model is usually specific to a business need and describes specific protocols.

Author: Yes, thanks.

> Each model is usually specific to a business need and describes specific protocols.

Author: Addressed by https://github.com/digital-asset/daml/pull/2505/commits/b43a821c448733c899f2e565e3654e0c63b37bca

Mocking a ledger response is usually not desirable to test the business logic, because so much of it is encapsulated in the DAML model.

This makes integration testing with an actual running ledger fundamental to evaluating the correctness of an application.

This is usually achieved by running a ledger as part of the test process and run several tests against it, possibly coordinated by a test framework.

Since the in-memory sandbox shipped as part of the SDK is a full-fledged implementation of a DAML ledger, it's usually the tool of choice for these tests.

Please note that this does not replace acceptance tests with the actual ledger implementation that your application aims to use in production.

Whatever your choice is, sharing a single ledger to run several tests is a suggested best practice.

> Share the ledger

Reviewer: This is too much text for the point you are making. Maybe combine with the next section and just add a single paragraph there to motivate sharing the ledger.

Author: Addressed by https://github.com/digital-asset/daml/pull/2505/commits/319dad7e7f104cce9f6556e5c93247d4656a7556

****************

Booting a ledger and loading DAML code into it takes time. The shorter your tests, the more this overhead is going to impact the running time of the tests.

Given the fundamental need of working against a running ledger to verify the business logic, it's quite common to have integration test that, in appearance, resemble unit tests, in that they check an individual property of a component you wrote, although in an integration test setting.

This means it's quite common to write short tests, which aim, just like unit tests, to give quick feedback to the developer.

In order for this to happen, saving on the need to boot a ledger multiple times can dramatically decrease the time necessary to run a suite of tests.

Since your tests are going to share the same ledger, they must be designed to not interfere with each other.

Design your tests to share the ledger
*************************************

> Both the transaction and the active contract service offer the possibility of filtering by party. Parties can thus be used as a way to isolate tests.

Reviewer: I think using disjoint sets of parties for each test is exactly the right approach, that's how Truffle does it.
However, your description here relies on starting with no active contracts. Getting the ledger into a desired state using automation is quite cumbersome. I often use scenarios. I think it'd be good to give some advice on ledger initialization.

Author: Scenarios can run only on a clean ledger, which would require to either disqualify the whole idea of sharing a ledger or need to run with a scenario that loads all possible events. Furthermore, this approach is incompatible with the suggested approach of allocating new parties for each test. Rather, one thing I left out and that greatly improves setting up the desired state is leveraging codegen where possible, which I could add as a suggestion.

Reviewer: Yeah, with scenarios you'd have to define your disjoint sets of parties in each scenario. Scenarios are composable so you can just write a container-scenario that runs the setup for each test .

Author: The problem in using this approach is having to manually keep synchronized an ever-growing list of users that have to carefully be created on both sides for this to work. I would favor using the `PartyManagementService` along with good helpers and possibly codegen to make the setup as lightweight as possible. This approach has been taken when developing the new integration tests and setting up a test scenario is very much not cumbersome ([example](https://github.com/digital-asset/daml/blob/master/ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/Divulgence.scala#L18-L22)). I can include your suggestion but it's important to highlight the possible maintenance burden for large test suites.

Author: After a bit of thought and interesting input from @gerolf-da, it's probably the case to reinforce the idea of having expressive bindings and possibly code generation do the heavy lifting for you of coming to a state.

In summary:

1. retrieve the current offset of the ledger end before the test starts
1. use the party management service to allocate the parties needed by the test
1. whenever you issue a command, issue it as one of the parties allocated for this test
1. whenever you need to get the set of active contracts or a stream of transactions, always filter by one or more of the parties allocated for this test

This isolation between instances of tests also means that different tests can be run completely in parallel with respect to each other, possibly improving on the overall running time of your test suite.

Reset if you need to
********************

It may be the case that you are running a very high number of tests, verifying the ins and outs of a very complex application interacting with an equally complex DAML model.

If that's the case, the leak of resources caused by the approach to test isolation mentioned above can become counterproductive, causing slow-downs or even crashes as the ledger backing your test suite has to keep track of more parties and more transactions that are actually no longer relevant after the test itself finishes.

> As a last resort for these cases, your tests can use a service that ledger implementations can optionally expose, designed exclusively for testing environments: the reset service.

Reviewer: ```suggestion
As a last resort for these cases, your tests can use the reset service, which ledger implementations can optionally expose for testing.
```
